src/renderer/wg_engine/tvgWgGeometry.h

```diff
@@ -24,102 +24,433 @@
+        for (uint32_t i = 0; i < buff.vcount; i++ ) {
+            vbuff[vcount + i] = buff.vbuff[i] + offset;
+            ibuff[icount + i] = vcount + i;
+        }
+        vcount += buff.vcount;
+        icount += buff.vcount;
+    };
+
+    // append line
+    inline void appendLine(const Point& v0, const Point& v1, float dist, float halfWidth) {
+        Point sub = v1 - v0;
+        Point nrm = { +sub.y / dist * halfWidth, -sub.x / dist * halfWidth };
+        appendQuad(v0 - nrm, v0 + nrm, v1 - nrm, v1 + nrm);
+    }
+
+    // append bevel joint
+    inline void appendBevel(const Point& v0, const Point& v1, const Point& v2, float dist1, float dist2, float halfWidth) {
+        Point sub1 = v1 - v0;
+        Point sub2 = v2 - v1;
+        Point nrm1 { +sub1.y / dist1 * halfWidth, -sub1.x / dist1 * halfWidth };
+        Point nrm2 { +sub2.y / dist2 * halfWidth, -sub2.x / dist2 * halfWidth };
+        appendQuad(v1 - nrm1, v1 + nrm1, v1 - nrm2, v1 + nrm2);
+    }
+
+    // append miter joint
+    inline void appendMitter(const Point& v0, const Point& v1, const Point& v2, float dist1, float dist2, float halfWidth, float miterLimit) {
+        Point sub1 = v1 - v0;
+        Point sub2 = v2 - v1;
+        Point nrm1 { +sub1.y / dist1, -sub1.x / dist1 };
+        Point nrm2 { +sub2.y / dist2, -sub2.x / dist2 };
+        Point offset1 = nrm1 * halfWidth;
+        Point offset2 = nrm2 * halfWidth;
+        Point nrm = normalize(nrm1 + nrm2);
+        float cosine = dot(nrm, nrm1);
+        float angle = std::acos(dot(nrm1, -nrm2));
+        float miterRatio = 1.0f / (std::sin(angle) * 0.5f);
+        if (miterRatio <= miterLimit) {
+            appendQuad(v1 + nrm * (halfWidth / cosine), v1 + offset2, v1 + offset1, v1);
+            appendQuad(v1 - nrm * (halfWidth / cosine), v1 - offset2, v1 - offset1, v1);
+        } else {
+            appendQuad(v1 - offset1, v1 + offset2, v1 - offset2, v1 + offset1);
+        }
+    }
+
+    // append square cap
+    inline void appendSquare(Point v0, Point v1, float dist, float halfWidth) {
+        Point sub = v1 - v0;
+        Point offset = sub / dist * halfWidth;
+        Point nrm = { +offset.y, -offset.x };
+        appendQuad(v1 - nrm, v1 + nrm, v1 + offset - nrm, v1 + offset + nrm);
+    }
+
+    // append strokes
+    inline void appendStrokes(const WgVertexBuffer& buff, const RenderStroke* rstroke) {
+        assert(rstroke);
+        // empty buffer gueard
+        if (buff.vcount < 2) return;
+        float halfWidth = rstroke->width * 0.5f;
+
+        // append core lines
+        for (size_t i = 1; i < buff.vcount; i++)
+            appendLine(buff.vbuff[i-1], buff.vbuff[i], buff.vdist[i], halfWidth);
 
-struct WgGeometryData
-{
-    static WgMath* gMath;
+        // append caps (square)
+        if ((rstroke->cap == StrokeCap::Square) && !buff.closed) {
+            appendSquare(buff.vbuff[0], buff.vbuff[1], buff.vdist[1], halfWidth);
```

Comment:
```appendSquare(buff.vbuff[1], buff.vbuff[0], buff.vdist[1], halfWidth);```
the first one has to be drawn backward. 
this why the first cap was not present (Stroke.cpp):
before:
<img width="138" alt="Zrzut ekranu 2024-09-24 o 12 28 47" src="https://github.com/user-attachments/assets/546c8304-ea2b-4933-bf7f-8c933501b452">

after:
<img width="141" alt="Zrzut ekranu 2024-09-24 o 12 28 32" src="https://github.com/user-attachments/assets/aac27cc0-6139-4bac-901e-d0b78396e437">